### PR TITLE
Add example code for generating structuring elements.

### LINKED ISSUE #4431: Add examples to docstring for structuring elements to visualize the shape
## Description

Add examples to show for each of these functions:

* [ ] skimage.morphology.square(width[, dtype])
* [ ] skimage.morphology.rectangle(width, height)
* [ ] skimage.morphology.diamond(radius[, dtype])
* [ ] skimage.morphology.disk(radius[, dtype])
* [ ] skimage.morphology.cube(width[, dtype])
* [ ] skimage.morphology.octahedron(radius[, dtype])
* [ ] skimage.morphology.ball(radius[, dtype])
* [ ] skimage.morphology.octagon(m, n[, dtype])
* [ ] skimage.morphology.star(a[, dtype])

the typical shape of the structuring element.

Cc ( @emmanuelle : another first issue idea).

### PR DESCRIPTION
**Pull request by b-z**
## Description

Closes #4431 .
The example shows how to use functions in `skimage.morphology` to generate structuring elements.
The title of each plot indicates the call of the function.

Figure_1

## Checklist

- [ ] ~~[Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)~~
- [ ] ~~Gallery example in `./doc/examples` (new features only)~~
- [ ] ~~Benchmark in `./benchmarks`, if your changes aren't covered by an
  existing benchmark~~
- [ ] ~~Unit tests~~
- [x] Clean style in [the spirit of PEP8](https://www.python.org/dev/peps/pep-0008/)

## For reviewers

- [ ] Check that the PR title is short, concise, and will make sense 1 year
  later.
- [ ] Check that new functions are imported in corresponding `__init__.py`.
- [ ] Check that new features, API changes, and deprecations are mentioned in
      `doc/release/release_dev.rst`.
- [ ] Consider backporting the PR with `@meeseeksdev backport to v0.14.x`

### CODE REVIEW

File: doc/examples/numpy_operations/plot_structuring_elements.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
+"""
+=============================
+Generate structuring elements
+=============================
+
+This example shows how to use functions in :py:module:`skimage.morphology`
+to generate structuring elements.
```
Comment (sciunto) [inline doc/examples/numpy_operations/plot_structuring_elements.py]: I think this paragraph could be improved by explaining what is a structuring element and providing few examples of functions using them (like erosion, dilation, functions in rank filters)...

File: doc/examples/numpy_operations/plot_structuring_elements.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+# Visualize the elements.
+fig = plt.figure(figsize=(10, 10))
+
+idx = 1
+for title, struc in struc_2d:
+    ax = fig.add_subplot(3, 3, idx)
+    ax.imshow(struc, cmap="Paired", vmin=0, vmax=12)
+    for i in range(struc.shape[0]):
+        for j in range(struc.shape[1]):
```
Comment (sciunto) [inline doc/examples/numpy_operations/plot_structuring_elements.py]: Here we could use itertools.product to avoid nested loops, but it's optional.

**Comment by soupault** [comment]
Thanks for the PR! This looks great!

I think for 2D we should try keeping the same absolute scale of the pixels across the plots. For 3D, one of these visualization methods may suit better - https://matplotlib.org/3.2.0/gallery/mplot3d/voxels.html, https://matplotlib.org/3.2.0/gallery/mplot3d/3d_bars.html .

**Comment by b-z** [comment]
Thanks for the comments. I will improve these in my free time.

**Comment by sciunto** [comment]
@b-z Do you have a chance to work on this again?

@scikit-image/core As we will improve our API in the future (https://github.com/scikit-image/scikit-image/issues/4536), I propose to merge this PR soon at it is already good. We will have the occasion to revisit the example later on.

File: doc/examples/numpy_operations/plot_structuring_elements.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+# Generate 2D and 3D structuring elements.
+# [(title, data)]
+struc_2d = [
+    ("square(5)", square(5)),
+    ("rectangle(5, 3)", rectangle(5, 3)),
+    ("diamond(5)", diamond(5)),
+    ("disk(7)", disk(7)),
+    ("octagon(4, 7)", octagon(4, 7)),
+    ("star(6)", star(6))
+]
```
Comment (rfezzani) [inline doc/examples/numpy_operations/plot_structuring_elements.py:28]:
```suggestion
struc_2d = {
    "square(5)": square(5),
    "rectangle(5, 3)": rectangle(5, 3),
    "diamond(5)": diamond(5),
    "disk(7)": disk(7),
    "octagon(4, 7)": octagon(4, 7),
    "star(6)": star(6)
}
```

This is optional, but you may then use then `for title, struc in struc_2D.item()` in the following loop.

File: doc/examples/numpy_operations/plot_structuring_elements.py [hunk 4]
```diff
@@ -27,17 +27,18 @@
...
+    "ball(5)": ball(5)
+}
 
 # Visualize the elements.
 fig = plt.figure(figsize=(8, 8))
 
 idx = 1
-for title, struc in struc_2d:
+for title in struc_2d:
+    struc = struc_2d[title]
```
Comment (rfezzani) [inline doc/examples/numpy_operations/plot_structuring_elements.py:41]:
```suggestion
for title, struc in struc_2d.items():
```
;-)

File: doc/examples/numpy_operations/plot_structuring_elements.py [hunk 5]
```diff
@@ -47,7 +48,8 @@
     ax.set_title(title)
     idx += 1
 
-for title, struc in struc_3d:
+for title in struc_3d:
+    struc = struc_3d[title]
```
Comment (rfezzani) [inline doc/examples/numpy_operations/plot_structuring_elements.py:52]:
```suggestion
for title, struc in struc_3d.items():
```

**Comment by b-z** [comment]
> Thanks for the PR! This looks great!
> 
> I think for 2D we should try keeping the same absolute scale of the pixels across the plots. For 3D, one of these visualization methods may suit better - https://matplotlib.org/3.2.0/gallery/mplot3d/voxels.html, https://matplotlib.org/3.2.0/gallery/mplot3d/3d_bars.html .

Hello @soupault , thanks for your advice! 
I tried to change them to same scale. As for the 3D visualization method proposed, I prefer the voxel plot, and also used same scale.
Do you like the following result better?
image

File: doc/examples/numpy_operations/plot_structuring_elements.py [hunk 6]
```diff
@@ -0,0 +1,60 @@
+"""
+=============================
+Generate structuring elements
+=============================
+
+This example shows how to use functions in :py:module:`skimage.morphology`
```
Comment (rfezzani) [inline doc/examples/numpy_operations/plot_structuring_elements.py:6]: Try this to fix the CI failure
```suggestion
This example shows how to use functions in :py:mod:`skimage.morphology`
```

**Comment by jni** [comment]
@b-z I love your latest plots! Could you push the code for those to your branch? It is still showing the dots and the different element sizes.

I also like @rfezzani's suggestions for iterating over the values. However, since you are not going to access the dictionary, it might be better to revert back to a list of tuples! Sorry about the back and forth there...!

Anyway, all that is optional, but the shaded voxels look super nice and we should get that code in since you worked it out!

**Comment by rfezzani** [comment]
> However, since you are not going to access the dictionary,

I did not get the point. Sorry @jni, can you elaborate more on this? :slightly_smiling_face:

**Comment by sciunto** [comment]
Me too. @rfezzani suggestion looks appropriate to me.

**Comment by jni** [comment]
@rfezzani 

Currently we are doing:

```python
structs2d = {'square(5)': square(5), ...}
...
for title, struc in structs2d.items():
    ...
```

I suggest:

```python
structs2d = [('square(5)', square(5)), ...]
...
for title, struc in structs2d:
    ...
```

**Comment by rfezzani** [comment]
I understood that @jni, I just still don't get your objection:
> However, since you are not going to access the dictionary,

If it is simply a matter of code style, no problem... Everybody has its own preferences ;-)

**Comment by b-z** [comment]
> @rfezzani
> 
> Currently we are doing:
> 
> ```python
> structs2d = {'square(5)': square(5), ...}
> ...
> for title, struc in structs2d.items():
>     ...
> ```
> 
> I suggest:
> 
> ```python
> structs2d = [('square(5)', square(5)), ...]
> ...
> for title, struc in structs2d:
>     ...
> ```

The newest code is using the style rfezzani suggested.
I'm new to open source communities, and cannot decide which to use. 
Thanks 😃

**Comment by jni** [comment]
@rfezzani what I meant is that we are not using the mapping feature of the dictionary. We stuff things into a dictionary then immediately unpack it into tuples. So we might as well just use the tuples.

@b-z Don’t worry, this is indeed just a suggestion. It’s common for maintainers to have disagreements, and it’s something you get used to over time — you eventually develop a sense for what are requirements and what are minor suggestions. Either way is minor with this dict/tuple question. Thank you so much for your work!

File: doc/examples/numpy_operations/plot_structuring_elements.py (hunk 1 above)
Comment (b-z) [reply doc/examples/numpy_operations/plot_structuring_elements.py:7]: Hello sciunto. Thank you for pointing it out : )
I felt a bit hard to explain what structuring element is (my English is bad..).
I made this as an example, which modified from the current example of `dilation` function:
```python
>>> import numpy as np
>>> from skimage.morphology import dilation, square, disk
>>> bright_pixel = np.array([[0, 0, 0, 0, 0],
...                          [0, 0, 0, 0, 0],
...                          [0, 0, 1, 0, 0],
...                          [0, 0, 0, 0, 0],
...                          [0, 0, 0, 0, 0]], dtype=np.uint8)
>>> dilation(bright_pixel, square(3))
array([[0, 0, 0, 0, 0],
       [0, 1, 1, 1, 0],
       [0, 1, 1, 1, 0],
       [0, 1, 1, 1, 0],
       [0, 0, 0, 0, 0]], dtype=uint8)
>>> dilation(bright_pixel, disk(2))
array([[0, 0, 1, 0, 0],
       [0, 1, 1, 1, 0],
       [1, 1, 1, 1, 1],
       [0, 1, 1, 1, 0],
       [0, 0, 1, 0, 0]], dtype=uint8)
```
Not sure if it should be put into the code.

**Comment by soupault** [comment]
@b-z thanks a lot for the updates! This looks amazing - https://github.com/scikit-image/scikit-image/pull/4528#issuecomment-616169102! So happy to have this work merged!

**Pull request merged**